### ISSUE TITLE
Unnecessary Operator triggered on short circuit evaluation

### ISSUE THREAD

**Issue by W4RH4WK** [Manishearth/rust-clippy#1194]
Hey there, I recently encountered a (what I think) false positive:

```
warning: statement can be reduced, #[warn(unnecessary_operation)] on by default
   --> src/front/ast.rs:232:21
    |
232 |                     arg.visit_expr(do_expr) || return false;
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
help: replace it with
    |                     arg.visit_expr(do_expr);return false;
    = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#unnecessary_operation
```

`arg.visit_expr(..)` returns a boolean which I use to determine whether to exit from the current function or not. This code should be equivalent to:

``` rust
if !arg.visit_expr(do_expr) {
    reutrn false;
}
```

Rewriting it to `arg.visit_expr(do_expr);return false;` would change the semantic of the program.

I know using the `if` would be more readable in general, but I'd like to keep these parts of the function more concise.

**Comment by llogiq** [comment]
I sort of disagree with your reasoning (keeping it short should not impede readability), but I'm pretty opinionated when it comes to code style, so take my judgement with a helping of salt.

That said, we should probably skip the lint if the statement has a return statement unless said statement is in a closure or embedded fn item. Should be doable with a custom `Visitor`.

**Comment by mcarton** [comment]
`foo || return bar` certainly is not idiomatic rust.

**Comment by W4RH4WK** [comment]
> foo || return bar certainly is not idiomatic rust.

I agree with you and I'll probably change this in the near future.

**Comment by oli-obk** [comment]
How about a lint forbidding diverging expressions except in match arms and as statements?

**Comment by mcarton** [comment]
@oli-obk :+1:

**Comment by samueltardieu** [comment]
Note that this is not specific to return. While writing some pathfinding function, I happened to have to remove an element from one vector, unless it was not found in which case I have to remove it from a second vector.

``` rust
  // Return true if the element has been found and removed from the vector
  fn remove_element<T: Eq>(v: &mut Vec<T>, e: &T) -> bool { … }

  // Used as:
  remove_element(&mut vec1, &e) || remove_element(&mut vec2, &e);
```

Here, Clippy suggested that I use ";" between the `remove_element` calls, which would not make it equivalent.

**Comment by oli-obk** [comment]
Indeed the suggestion is wrong. Anyone who wants to fix this bug, simply change the suggestion for short circuiting ops used like that to an `if !a { b; }` or not negated depending on whether it's `&&` or `||`
